### PR TITLE
AVM: Go19 curve check

### PR DESCRIPTION
**Pull request by jannotti**
Go 1.19 has changed the semantics of ecdsa.Verify to panic if the public key is not on the appropriate curve.  This change prepares for supporting Go 1.19 by first establishing a new rule for the AVM opcode ecdsa_verify. By doing this in a consensus upgrade while still using Go 1.17, we force the change to occur across all nodes at once, and we ensure no changes occur in evaluating old blocks.

### CODE REVIEW

**Review by algorandskiy — DISMISSED** [review]

**Review by ahangsu — COMMENTED** [review]
Generally looks right to me, I agree with Idan's idea of saving one consensus param by `&& secp256r1.IsOnCurve(x, y)`, such that we can ensure the pk to be on curve without panicing in verify.

Only one question for Idan @algoidan 
> Do we suspect that there will be a valid signature the will fail on IsOnCurve?

Do you mean we want to check the signature on curve? (I think signature elements should be?)

**Comment by bbroder-algo** [comment]
looks good to me, I see you swapped the calls to SetBytes and PublicKey but they look independent also now its in the order of the argument list for Verify

**Comment by jannotti** [comment]
> looks good to me, I see you swapped the calls to SetBytes and PublicKey but they look independent also now its in the order of the argument list for Verify
> 

Now that there's a potential early failure for being off curve, I figured I'd defer the work to create the big ints unless the check passes.

File: data/transactions/logic/eval.go [hunk 1]
```diff
@@ -3421,15 +3423,16 @@ func opEcdsaVerify(cx *EvalContext) error {
...
 		result = secp256k1.VerifySignature(pubkey, msg, signature)
 	} else if fs.field == Secp256r1 {
-		r := new(big.Int).SetBytes(sigR)
-		s := new(big.Int).SetBytes(sigS)
-
-		pubkey := ecdsa.PublicKey{
-			Curve: elliptic.P256(),
-			X:     x,
-			Y:     y,
+		if !cx.Proto.EnablePrecheckECDSACurve || secp256r1.IsOnCurve(x, y) {
```
Comment (brianolson) [inline data/transactions/logic/eval.go:3426]: `if !IsOnCurve()` shouldn't that be an error? Or is it just `result = false` not verified?

Comment (jannotti) [reply data/transactions/logic/eval.go:3426]: I think there's an argument for making it an error, but we did not before, and I wanted the change to be as small as possible. Do you think there are use cases that need to distinguish an error like this from failing to verify?

Comment (ahangsu) [reply data/transactions/logic/eval.go:3426]: I will think it is somewhat sus to make a different case of erroring and failing to verify... feeling like giving a larger attack surface against verifier?

https://www.reddit.com/r/cryptography/comments/qyg723/what_happens_when_a_public_key_point_does_not_lie/ Makes me recall something like verifier rejection attack (replay attack) and invalid curve attack...

Comment (bbroder-algo) [reply data/transactions/logic/eval.go:3426]: I think this patch improves the code.

**Review by bbroder-algo — APPROVED** [review]
Returns false on verifications before performing ECDSA verification for Secp256r1 curves if x, y components of public keys are not on the curve to begin with.

**Pull request merged**